src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java

```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
-                           .getResultSetMetaData().getSearchParameters()));
-                   break;
-               }
-               
-               // All permission checks must pass
-               filteringResultSet.setIncluded(i, true);
-   
-               NodeRef nodeRef = returnedObject.getNodeRef(i);
-               
-               if(filteringResultSet.getIncluded(i) && (nodeRef == null))
-               {
-                   filteringResultSet.setIncluded(i, false);
-               }
-   
-               if (filteringResultSet.getIncluded(i) && permissionService.hasReadPermission(nodeRef) == AccessStatus.DENIED)
-               {
-                   filteringResultSet.setIncluded(i, false);
-               }
-   
-               // Bug out if we are limiting by size
-               if ((maxSize != null) && (filteringResultSet.length() > maxSize.intValue()))
-               {
-                   // Renove the last match to fix the correct size
-                   filteringResultSet.setIncluded(i, false);
-                   filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.FINAL_SIZE, PermissionEvaluationMode.EAGER, returnedObject.getResultSetMetaData()
-                           .getSearchParameters()));
-                   break;
-               }
-           }
+            for (int i = 0; i < returnedObject.length(); i++)
+            {
+                long currentTimeMillis = System.currentTimeMillis();
+
+                NodeRef nodeRef = returnedObject.getNodeRef(i);
+
+                // All permission checks must pass
+                filteringResultSet.setIncluded(i, true);
+
+                if (i >= maxChecks)
+                {
+                    log.warn("maxChecks exceeded (" + maxChecks + ")", new Exception("Back Trace"));
+                    filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.NUMBER_OF_PERMISSION_EVALUATIONS, PermissionEvaluationMode.EAGER, returnedObject
+                            .getResultSetMetaData().getSearchParameters()));
+                    break;
+                }
+                else if ((currentTimeMillis - startTimeMillis) > maxCheckTime)
+                {
+                    log.warn("maxCheckTime exceeded (" + (currentTimeMillis - startTimeMillis) + " milliseconds)", new Exception("Back Trace"));
+                    filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.NUMBER_OF_PERMISSION_EVALUATIONS, PermissionEvaluationMode.EAGER, returnedObject
+                            .getResultSetMetaData().getSearchParameters()));
+                    break;
+                }
+
+                // If supportedDefinitions is not passed as parameter, it uses new method to permission filtering.
+                if (supportedDefinitions == null)
+                {
+                    if (permissionService.hasReadPermission(nodeRef) == AccessStatus.DENIED)

```

Comment:
These two statements could be AND-ed together.

Reply:
Oh - actually I didn't spot the `else` statement.  Instead we could invert the outside if statement to get:
```
if (supportedDefinitions != null)
{
    ...
}
else if (permissionService.hasReadPermission(nodeRef) == DENIED)
{
    ...
}
```